### PR TITLE
8297878: KEM: Implementation

### PR DESCRIPTION
**Pull request by wangweij**
This is the KEM API backport to Java SE 17 MR 1.

The `src` files are identical to those in the current `jdk` repo except for the change made to `KEM.java` at https://github.com/openjdk/jdk/commit/59c2aff1edffb66762bbbe5e310913f87953be5b#diff-7bee547996d0de5692181a509bdf509276c7eb9351722580fd6aee7975745e67.

_Update: There are javadoc changes to all the `src` files in following commits._

The `RSA_KEM` code is modified because `DerOutputStream` lacks several methods in JDK 17.

`Proc` is updated like in JDK 21 to support for some internal interop testing. The test files for `Proc` are also backported.

This change does not contain the DHKEM implementation in the original JDK 21 change.

This change also covers JDK-8322971 which fixed a follow-on P3 bug of the initial KEM work.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] [JDK-8322971](https://bugs.openjdk.org/browse/JDK-8322971) needs maintainer approval
- [x] Commit message must refer to an issue
- [x] [JDK-8297878](https://bugs.openjdk.org/browse/JDK-8297878) needs maintainer approval
- [x] Change requires CSR request [JDK-8330545](https://bugs.openjdk.org/browse/JDK-8330545) to be approved

### Issues
 * [JDK-8297878](https://bugs.openjdk.org/browse/JDK-8297878): KEM: Implementation (**Enhancement** - P3 - Approved) ⚠️ Issue is already resolved. Consider making this a "backport pull request" by setting the PR title to `Backport <hash>` with the hash of the original commit. See [Backports](https://wiki.openjdk.org/display/SKARA/Backports).
 * [JDK-8322971](https://bugs.openjdk.org/browse/JDK-8322971): KEM.getInstance() should check if a 3rd-party security provider is signed (**Bug** - P3 - Approved) ⚠️ Issue is already resolved. Consider making this a "backport pull request" by setting the PR title to `Backport <hash>` with the hash of the original commit. See [Backports](https://wiki.openjdk.org/display/SKARA/Backports).
 * [JDK-8330545](https://bugs.openjdk.org/browse/JDK-8330545): KEM: Implementation (**CSR**)

### Reviewers
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**)
 * [Iris Clark](https://openjdk.org/census#iris) (@irisclark - **Reviewer**)
 * [Andrew John Hughes](https://openjdk.org/census#andrew) (@gnu-andrew - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk17u-ri.git pull/1/head:pull/1` \
`$ git checkout pull/1`

Update a local copy of the PR: \
`$ git checkout pull/1` \
`$ git pull https://git.openjdk.org/jdk17u-ri.git pull/1/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1`

View PR using the GUI difftool: \
`$ git pr show -t 1`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk17u-ri/pull/1.diff">https://git.openjdk.org/jdk17u-ri/pull/1.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk17u-ri/pull/1#issuecomment-2064079468)

### CODE REVIEW

File: src/java.base/share/classes/javax/crypto/KEMSpi.java [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+        int engineEncapsulationSize();
+    }
+
+    /**
+     * The KEM decapsulator implementation, generated by
+     * {@link #engineNewDecapsulator} on the KEM receiver side.
+     *
+     * @see KEM.Decapsulator
+     *
+     * @since 21
```
Comment (seanjmullan) [inline src/java.base/share/classes/javax/crypto/KEMSpi.java]: Change the `@since` tag.

Comment (wangweij) [reply src/java.base/share/classes/javax/crypto/KEMSpi.java]: Oh, I didn't notice these inner ones. Will do. Thanks.

**Review by seanjmullan — APPROVED** [review]

**Review by irisclark — APPROVED** [review]
The Spec changes appear to be identical to those in JDK 21, with the exception of the expected `@since 17` and `@apiNote` which are used to identify these as additions via MR.

Associated CSR (JDK-8330545) also Reviewed.

**Review by gnu-andrew — APPROVED** [review]
It would be easier to review if [JDK-8322971](https://bugs.openjdk.org/browse/JDK-8322971) and [JDK-8305846](https://bugs.openjdk.org/browse/JDK-8305846): "Support compilation in Proc test utility" were backported separately.  Also, 8322971 seems to be missing from 21u as well.

The changes look correct as far as I can tell. I'm not sure why different copyright years are being used from the original patch, particularly for `Provider.java` where this is the only difference from 8297878.

The other differences (21->17, removal of the implementation, SSL & ECC changes) make sense. I presume the test differences with `getKemImpl` are due to the implementation removal making `KEM.getInstance` unusable?

So Skara correctly recognises this as a backport, the title should be "Backport 6b90b0519e89429300838fa598b2ea9ffda984a2"

**Comment by GoeLin** [comment]
Hi @wangweij, you might want to enable the GHA tests?

**Comment by wangweij** [comment]
@gnu-andrew I backported the 3 issues in a single PR because [JDK-8305846](https://bugs.openjdk.org/browse/JDK-8305846) was originally created to support the RSA-KEM interop test. I understand the change for it looks a little strange without any usage of it here. I hope grouping them together will make further backport - if exists - simpler.

For the year of `Provider.java`, I just updated it to *now*.

Yes, `getKemImpl` exists because after JDK-8322971 one cannot load a KEM impl from an unsigned security provider in Oracle JDK. This is a little different from OpenJDK.

**Comment by wangweij** [comment]
@GoeLin Maybe you can ask the maintainer of this repo to decide whether GHA tests can be enabled. Some security-related tests would fail because hardcoded certificates have expired.

**Comment by gnu-andrew** [comment]
> 8322971



> @gnu-andrew I backported the 3 issues in a single PR because [JDK-8305846](https://bugs.openjdk.org/browse/JDK-8305846) was originally created to support the RSA-KEM interop test. I understand the change for it looks a little strange without any usage of it here. I hope grouping them together will make further backport - if exists - simpler.
> 

We usually try and avoid it for more complex patches. It makes it easier for the person backporting, but harder for reviewers who have to pick it apart again.

> For the year of `Provider.java`, I just updated it to _now_.

Ok, we usually avoid this as it creates differences from the original commit which may then break further backports.

> 
> Yes, `getKemImpl` exists because after JDK-8322971 one cannot load a KEM impl from an unsigned security provider in Oracle JDK. This is a little different from OpenJDK.

Ah, ok, so it's  necessary after 8322971.

**Pull request closed without merging**